clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java

```diff
@@ -60,6 +61,49 @@ public void partitionAssignment() {
     }
 
     @Test
+    public void partitionAssignmentOnSubscriptionChange() {
+        state.assignFromUser(new HashSet<>(Arrays.asList(tp0, tp1)));
+        // assigned partitions should immediately change
+        assertEquals(2, state.assignedPartitions().size());
+        assertTrue(state.assignedPartitions().contains(tp0));
+        assertTrue(state.assignedPartitions().contains(tp1));
+
+        state.unsubscribe();
+        // assigned partitions should immediately change
+        assertTrue(state.assignedPartitions().isEmpty());
+
+
+        state.subscribe(singleton(topic1), rebalanceListener);
+        // assigned partitions should remain unchanged
+        assertTrue(state.assignedPartitions().isEmpty());
+
+        state.assignFromSubscribed(asList(t1p0));
+        // assigned partitions should immediately change
+        assertEquals(singleton(t1p0), state.assignedPartitions());
+
+        state.unsubscribe();
+        // assigned partitions should immediately change
+        assertTrue(state.assignedPartitions().isEmpty());
+
+
+        state.subscribe(Pattern.compile(".*"), rebalanceListener);
```

Comment:
nit: would you mind splitting from here on into a separate test case specific to patterns? The method name serves as documentation and also isolates the functionality under test a little better which makes debugging easier.